src/t_hash.c

```diff
@@ -29,52 +29,206 @@
 
 #include "server.h"
 #include <math.h>
+#include <stdbool.h>
 
 /*-----------------------------------------------------------------------------
  * Hash Entry API
  *----------------------------------------------------------------------------*/
 
-struct hashTypeEntry {
+/* The hashTypeEntry pointer is the field sds. We encode the entry layout type
+ * in the field SDS header. Field type SDS_TYPE_5 doesn't have any spare bits to
+ * encode this so we use it only for the first layout type.
+ *
+ * ENTRY_ENC_EMB_VALUE, used when it fits in a cache line. The value is stored
+ * as SDS_TYPE_8. The field can use any SDS type.
+ *
+ *     +--------------+---------------+
+ *     | field        | value         |
+ *     | hdr "foo" \0 | hdr8 "bar" \0 |
+ *     +------^-------+---------------+
+ *            |
+ *            |
+ *          entry pointer = field sds
+ *
+ * ENTRY_ENC_PTR_VALUE, with value pointer, used for larger fields and values.
+ * The field is SDS type 8 or higher.
+ *
+ *     +-------+--------------+
+ *     | value | field        |
+ *     | ptr   | hdr "foo" \0 |
+ *     +-------+------^-------+
+ *                    |
+ *                    |
+ *                 entry pointer = field sds
+ */
+
+typedef enum {
+    ENTRY_ENC_EMB_VALUE = 0,
+    ENTRY_ENC_PTR_VALUE = 1
+} hashTypeEntryEnc;
+
+#define FIELD_SDS_AUX_BIT_ENTRY_HAS_PTR_VALUE 0
+
+/* Struct used for ENTRY_ENC_PTR_VALUE. */
+typedef struct {
     sds value;
-    char field_offset;
     char field_data[];
-};
+} hashTypeEntryPtrValue;
+
+static inline hashTypeEntryEnc entryGetEncoding(const hashTypeEntry *entry) {
+    return sdsGetAuxBit(entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_PTR_VALUE);
+}
+
+/* Returns the containing struct for an entry without embedded value. */
+static hashTypeEntryPtrValue *getEntryStruct(const hashTypeEntry *entry) {
+    serverAssert(entryGetEncoding(entry) == ENTRY_ENC_PTR_VALUE);
+    unsigned char *buf = sdsAllocPtr(entry);
+    buf -= offsetof(hashTypeEntryPtrValue, field_data);
+    return (void *)buf;
+}
```

Comment:
I understand the motivation but IMO we should use a more explicit name for this:

```suggestion
static inline sds *encodedPtrValueEntryGetValueref(const hashTypeEntry *entry) {
    serverAssert(entryGetEncoding(entry) == ENTRY_ENC_PTR_VALUE);
    return (sds *)(sdsAllocPtr((sds)entry)) - 1;
}
```

Reply:
Sure, if accessing the value is the only thing it's used for... I'll look tomorrow.